### PR TITLE
[AzCopyV10][StgExp] Set blob tags when the user explicitly specifies tags.

### PR DESCRIPTION
**Pull request by mohsha-msft**

### CODE REVIEW

File: cmd/copy.go [hunk 1]
```diff
@@ -650,10 +659,10 @@ func (raw rawCopyCmdArgs) cookWithId(jobId common.JobID) (cookedCopyCmdArgs, err
...
 		// For s3 and file, only hot block blob tier is supported.
-		if cooked.blockBlobTier != common.EBlockBlobTier.None() ||
-			cooked.pageBlobTier != common.EPageBlobTier.None() {
-			return cooked, fmt.Errorf("blob-tier is not supported while copying from service to service")
-		}
+		//if cooked.blockBlobTier != common.EBlockBlobTier.None() ||
+		//	cooked.pageBlobTier != common.EPageBlobTier.None() {
+		//	return cooked, fmt.Errorf("blob-tier is not supported while copying from service to service")
+		//}
 		if cooked.noGuessMimeType {
```
Comment (zezha-msft) [inline cmd/copy.go:666]: I'll ignore the tier changes in this PR.

File: common/fe-ste-models.go [hunk 2]
```diff
@@ -934,6 +938,11 @@ func (m Metadata) ToAzFileMetadata() azfile.Metadata {
 	return azfile.Metadata(m)
 }
 
+// ToAzBlobTagsMap converts BlobTagsMap to azblob's BlobTagsMap
+func (bt BlobTags) ToAzBlobTagsMap() azblob.BlobTagsMap {
+	return azblob.BlobTagsMap(bt)
+}
```
Comment (zezha-msft) [inline common/fe-ste-models.go:944]: please pull the BlobTags code together for readability.

File: common/fe-ste-models.go [hunk 3]
```diff
@@ -1018,6 +1032,27 @@ func (m Metadata) ExcludeInvalidKey() (retainedMetadata Metadata, excludedMetada
...
+	return strings.Join(lst, "&")
+}
+
+func ToCommonBlobTagsMap(blobTagsString string) BlobTags {
+	if blobTagsString == "" {
+		return nil
+	}
+
+	blobTagsMap := BlobTags{}
+	for _, keyAndValue := range strings.Split(blobTagsString, "&") { // key/value pairs are separated by ';'
```
Comment (zezha-msft) [inline common/fe-ste-models.go:1049]: >// key/value pairs are separated by ';' [](start = 66, length = 39)

fix comment

File: cmd/copy.go [hunk 4]
```diff
@@ -814,6 +823,53 @@ func validateMd5Option(option common.HashValidationOption, fromTo common.FromTo)
...
 
+// Valid tag key and value characters include:
+// 1. Lowercase and uppercase letters (a-z, A-Z)
+// 2. Digits (0-9)
+// 3. A space ( )
+// 4. Plus (+), minus (-), period (.), solidus (/), colon (:), equals (=), and underscore (_)
+func isValidBlobTagsKeyValue(keyVal string) bool {
+	for _, c := range keyVal {
+		if !((c >= '0' && c <= '9') || (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || c == ' ' || c == '+' ||
+			c == '-' || c == '.' || c == '/' || c == ':' || c == '=' || c == '_') {
```
Comment (zezha-msft) [inline cmd/copy.go:834]: Hmm I don't think this check is correct, the +-./:=_ are all query-encoded, no? ToCommonBlobTagsMap didn't decode them, right?

File: cmd/copyEnumeratorInit.go [hunk 5]
```diff
@@ -102,6 +102,10 @@ func (cca *cookedCopyCmdArgs) initEnumerator(jobPartOrder common.CopyJobPartOrde
 		cca.stripTopDir = true
 	}
 
+	if cca.fromTo.To() != common.ELocation.Blob() && cca.blobTags != nil {
+		return nil, fmt.Errorf("blob tags can only be set when transferring to blob storage")
+	}
```
Comment (zezha-msft) [inline cmd/copyEnumeratorInit.go:107]: this check fits better in the cook function

Comment (mohsha-msft) [reply cmd/copyEnumeratorInit.go:107]: Ohh I was wondering about that too. Sure, I'll move it.

File: common/fe-ste-models.go (hunk 3 above)
Comment (mohsha-msft) [reply common/fe-ste-models.go:1049]: Done

File: common/fe-ste-models.go (hunk 2 above)
Comment (mohsha-msft) [reply common/fe-ste-models.go:944]: Done

File: cmd/helpMessages.go [hunk 6]
```diff
@@ -79,6 +79,14 @@ Upload files and directories by using a SAS token and wildcard (*) characters:
 
   - azcopy cp "/path/*foo/*bar*" "https://[account].blob.core.windows.net/[container]/[path/to/directory]?[SAS]" --recursive=true
 
+Upload files and directories to Azure Storage account and set the query-string encoded tags on the blob. 
+
+	- To set tags {key = "bla bla", val = "foo"} and {key = "bla bla 2", val = "bar"}, use the following syntax :
+		- azcopy cp "/path/*foo/*bar*" "https://[account].blob.core.windows.net/[container]/[path/to/directory]?[SAS]" --blob-tags="bla%20bla=foo&bla%20bla%202=bar"
+	- Keys and values are URL encoded and the key-value pairs are separated by an ampersand('&')
+	- https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blob-index-how-to?tabs=azure-portal
+	- While setting tags on the blobs, there are additional permissions('t' for tags and 'f' for find blobs by tags) in SAS without with service with give AuthZ error.
```
Comment (zezha-msft) [inline cmd/helpMessages.go:88]: >with [](start = 129, length = 4)

typo: without which the service will give authorization error back.

Comment (zezha-msft) [inline cmd/helpMessages.go:88]: >'f' for find blobs by tag [](start = 86, length = 25)

maybe clarify that 'f' is not needed for AzCopy

File: ste/JobPartPlanFileName.go [hunk 7]
```diff
@@ -101,6 +101,18 @@ func (jpfn JobPartPlanFileName) Create(order common.CopyJobPartOrderRequest) {
...
+	if len(order.BlobAttributes.BlobTagsString) > len(JobPartPlanDstBlob{}.BlobTags) {
+		panic(fmt.Errorf("blob tags string is too large: %q", order.BlobAttributes.BlobTagsString))
+	}
+	/*
+	*	TODO: Remove this comment
+	*	Since I've already verified that at most 10 blob tags can be set and put a restriction on key and value,
+	*	there is no need to check the length. size of JobPartPlanDstBlob{}.BlobTags = 4kb
+	*	key(128) + value(256) + separator('&', 1) = 385
+	*	10 tags * 385 = 3850
+	*	+ 10 * characters(';', 1) for delimiter
```
Comment (zezha-msft) [inline ste/JobPartPlanFileName.go:113]: >; [](start = 22, length = 1)

minor: delimiter is =

File: ste/mgr-JobPartMgr.go [hunk 8]
```diff
@@ -341,6 +344,16 @@ func (jpm *jobPartMgr) ScheduleTransfers(jobCtx context.Context) {
 			jpm.metadata[kv[0]] = kv[1]
 		}
 	}
+	blobTagsStr := string(dstData.BlobTags[:dstData.BlobTagsLength])
+	jpm.blobTags = common.BlobTags{}
+	if len(blobTagsStr) > 0 {
+		for _, keyAndValue := range strings.Split(blobTagsStr, "&") { // key/value pairs are separated by ';'
+			kv := strings.Split(keyAndValue, "=") // key/value are separated by '&'
```
Comment (zezha-msft) [inline ste/mgr-JobPartMgr.go:351]: >'&' [](start = 71, length = 3)

minor fix comments

File: ste/sender-appendBlob.go [hunk 9]
```diff
@@ -139,11 +141,21 @@ func (s *appendBlobSenderBase) Prologue(ps common.PrologueState) (destinationMod
 	s.headersToApply.ContentType = ps.GetInferredContentType(s.jptm)
 
 	destinationModified = true
-	_, err := s.destAppendBlobURL.Create(s.jptm.Context(), s.headersToApply, s.metadataToApply, azblob.BlobAccessConditions{})
-	if err != nil {
+	blobTags := s.blobTagsToApply
+	setTagsRequired := setTagsRequired(blobTags)
```
Comment (zezha-msft) [inline ste/sender-appendBlob.go:145]: >setTagsRequired [](start = 1, length = 15)

Readability: consider rename to something like separateSetTagRequired to make its intent clearer

**Review by zezha-msft — APPROVED** [review]
:shipit:

File: ste/sender-appendBlob.go (hunk 9 above)
Comment (mohsha-msft) [reply ste/sender-appendBlob.go:145]: Done

File: ste/mgr-JobPartMgr.go (hunk 8 above)
Comment (mohsha-msft) [reply ste/mgr-JobPartMgr.go:351]: Done

File: ste/JobPartPlanFileName.go (hunk 7 above)
Comment (mohsha-msft) [reply ste/JobPartPlanFileName.go:113]: Done

File: cmd/helpMessages.go (hunk 6 above)
Comment (mohsha-msft) [reply cmd/helpMessages.go:88]: Done

Comment (mohsha-msft) [reply cmd/helpMessages.go:88]: Done

File: ste/sender.go [hunk 10]
```diff
@@ -203,3 +203,19 @@ func newBlobUploader(jptm IJobPartTransferMgr, destination string, p pipeline.Pi
 		return newBlockBlobUploader(jptm, destination, p, pacer, sip) // If no blob type was inferred, assume block blob.
 	}
 }
+
+const TagsHeaderMaxLength = 2000
+
+// If length of tags <= 2kb, pass it in the header x-ms-tags. Else do a separate SetTags call
+func separateSetTagsRequired(tagsMap azblob.BlobTagsMap) bool {
```
Comment (nakulkar-msft) [inline ste/sender.go:210]: This function should rather be method on azblob.BlobTagsMap

**Review by nakulkar-msft — APPROVED** [review]

**Pull request merged**